include/drivers/spi.h

```diff
@@ -376,17 +379,120 @@ struct spi_dt_spec {
 	SPI_DT_SPEC_GET(DT_DRV_INST(inst), operation_, delay_)
 #endif
 
+/**
+ * SPI Enhanced Mode flags: Double Data Rate, Dual/Quad/Octal modes
+ * (Respectively DDR and DQO)
+ */
+
+/**
+ * SPI Command phase is send via standard mode (single line)
+ * spi_config lines bit field will be superseded by this flag only
+ * for the command.
+ */
+#define SPI_EM_CMD_STD		BIT(0)
+
+/**
+ * SPI Command phase is send via enhanced modes (double/quad/octal)
+ * The mode will depend on the spi_config's lines bit field.
+ */
+#define SPI_EM_CMD_DQO		BIT(1)
+/**
+ * SPI Command phase will take advantage of Double Data Rate
+ * This flag is independent of SPI_EM_CMD_STD or SPI_EM_CMD_DQO.
+ */
+#define SPI_EM_CMD_DDR		BIT(2)
+/**
+ * SPI Addressing phase is send via standard mode (single line)
+ * spi_config's lines bit field will be superseded by this flag only
+ * for the address.
+ */
+#define SPI_EM_ADR_STD		BIT(3)
+/**
+ * SPI Addressing phase is send via enhanced modes (double/quad/octal)
+ * The mode will depend on the spi_config's lines bit field.
+ */
+#define SPI_EM_ADR_DQO		BIT(4)
```

Comment:
If I understand it correctly, these two flags are mutually exclusive. If so, why not use instead a 2-bit wide bit field that could take the values: `*_NONE` (no addressing phase) , `*_STD`, and `*_DQO`? Perhaps it would be good to also merge the bit field with the DDR flag (and use `*_STD_DDR` and `*_DQO_DDR` values), as the DDR flag alone has no point when there is no addressing phase.
Then the address length bit field could be 4-bit long (as the `*_NONE` value would be no longer needed there) and the documentation could specify that this field is ignored if there is no addressing phase.
Same scheme could be used for specification of the command phase (it could be then made possible to configure the command length to 12 bits; I don't know if such option is useful, though).

Reply:
To summarize:

- no addressing phase would be: !SPI_EM_ADR_STD && !SPI_EM_ADR_DQO 
- thus reducing address length to 4 bits (and default value would be then SPI_EM_ADR_LEN_4_BITS )
- instead of using SPI_EM_ADR_DDR directly, let's have dedicated SPI_EM_ADR_STD_DDR = (SPI_EM_ADR_STD | SPI_EM_ADR_DDR) and SPI_EM_ADR_DQO_DDR = (SPI_EM_ADR_DQO | SPI_EM_ADR_DDR) 
- same for command.

Makes sense

Reply:
While revising the API according to your comments, I found out I forgot one important config : the wait cycles (or dummies) after a transmission and right before a reception. This is mandatory for letting the slave to reconfigure itself properly.

And that would take more than the 2 left over bits I have. Looks like I'll need to raise flags attribute to 32bits. Note that this is a non issue: the compiler aligns the struct to 4bytes, so the size isn't going to change whether I use uint16_t or uint32_t for flags attribute. in the end spi_buf structure is going to get bloated by 4 bytes anyway.

Reply:
> If so, why not use instead a 2-bit wide bit field that could take the values: *_NONE (no addressing phase) , *_STD, and *_DQO? Perhaps it would be good to also merge the bit field with the DDR flag (and use *_STD_DDR and *_DQO_DDR values), as the DDR flag alone has no point when there is no addressing phase.

Actually, that's not possible: you still have the case where cmd and/or address are sent in enhanced modes (dual/quad/octal). Completely forgot about it: these are the defaults. So we need a way to tell if there is a cmd and/or address and the only way is through the length.